lib/task-graph.js

```diff
@@ -88,28 +132,22 @@ function taskGraphFactory(
         if (marker.permanentMark) {
             return;
         }
-
         marker.temporaryMark = true;
-
+        
         _.forEach(task.waitingOn, function(value, dep) {
             var nonTerminalOnStates = [];
-
-            // This won't actually get called because we do this same check in
-            // _populateTaskData, but keep this here just to be defensive.
-            if (!_.has(self.tasks, dep)) {
-                throw new Error('Graph does not contain task with ID ' + dep);
-            }
-            // Expand 'finished' to all finished task states
-            if (_.contains(value, Constants.Task.States.Finished)) {
-                value = Constants.Task.FinishedStates;
-                task.waitingOn[dep] = [Constants.Task.States.Finished];
+            if (dep === "either") {
```

Comment:
The mongodb will not allow user to define variable name that starts with '$', so you could first choose to add the assertion logic, I think this is one of the place that is suitable for you mentioned code: https://github.com/RackHD/on-tasks/blob/master/lib/task-graph.js#L344